gms/gossiper.cc

```diff
@@ -1708,13 +1712,58 @@ void gossiper::apply_new_states(inet_address addr, endpoint_state& local_state,
     utils::chunked_vector<application_state> changed;
     auto&& remote_map = remote_state.get_application_state_map();
 
+    auto revert_to_old_app_state = [&] {
+        try {
+            logger.warn("Started to roll back to old app state for node={}, old_state={}, status={}, replicate_failed={}, listener_failed={}",
+                    addr, old_state, get_gossip_status(addr), replicate_failed, listener_failed);
+            local_state = old_state;
+            replicate(addr, old_state).get();
+            logger.warn("Finished to roll back to old app state for node={}, old_state={}, status={}, replicate_failed={}, listener_failed={}",
+                    addr, old_state, get_gossip_status(addr), replicate_failed, listener_failed);
+        } catch (...) {
+            logger.warn("Failed to roll back to old app state for node={}, old_state={}, status={}: {}",
+                    addr, old_state, get_gossip_status(addr), std::current_exception());
+            bool critical_app_changed = false;
+            for (auto& app : changed) {
+                if (app == application_state::LOAD ||
+                    app == application_state::CACHE_HITRATES ||
+                    app == application_state::VIEW_BACKLOG) {
+                    continue;
+                }
+                critical_app_changed = true;
+            }
+            if (critical_app_changed) {
+                // This should be very rare. If this happens, we can do nothing
+                // but throw and abort scylla process to prevent to leave this
+                // node in a base state.
+                logger.error("Failed to roll back critical gossip app state for node={}, changed={}. Aborting scylla process to prevent leaving the node in an inconsistent state forerver ...",
+                              addr, changed);
+                throw;
+            } else {
+                logger.warn("Ignore non critical gossip app state for node={}, changed={}", addr, changed);
+            }
+        }
+    };
+
     // Exceptions thrown from listeners will result in abort because that could leave the node in a bad
     // state indefinitely. Unless the value changes again, we wouldn't retry notifications.
     // Some values are set only once, so listeners would never be re-run.
     // Listeners should decide which failures are non-fatal and swallow them.
```

Comment:
> Listeners should decide which failures are non-fatal and swallow them.

Why not use this approach?

Reply:
Because this approach does not work. With this approach, there are two cases

1) no error, happy path

2) there is  failure, the listener either throw (A: terminate whole scylla process) or swallow (B: leave the node in a bad state forever)

This patch introduce 3rd option C for case 2, that is allow gossip to run the listener again to recover. C is better than A and B.